### PR TITLE
Replace MD5 with SHA-256

### LINKED ISSUE #200: Replace MD5 with SHA-256
[MD5 is a weak hash known to have hash collisions.](https://github.com/MobSF/owasp-mstg/blob/master/Document/0x04g-Testing-Cryptography.md#identifying-insecure-andor-deprecated-cryptographic-algorithms-mstg-crypto-4)

Standards
```
CWE: CWE-327: Use of a Broken or Risky
Cryptographic Algorithm 
OWASP Top 10: M5: Insufficient Cryptography 
OWASP MASVS: MSTG-CRYPTO-4
```

Files
```
com/onaio/steps/helper/AuthDialog.java 
com/onaio/steps/utils/QRCodeUtils.java

```

### PR DESCRIPTION
**Pull request by owais-vd**
close #200

### CODE REVIEW

File: src/main/java/com/onaio/steps/helper/AuthDialog.java [hunk 1]
```diff
@@ -128,7 +128,7 @@ private void toggleFirstTimeEntry() {
     private void auth() {
         String enteredPW = passwordEditText.getText().toString();
```
Comment (Kigamba) [inline src/main/java/com/onaio/steps/helper/AuthDialog.java:129]: Check this for null `passwordEditText.getText()` can return null

File: src/main/java/com/onaio/steps/helper/HashGenerator.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+            StringBuilder hexString = new StringBuilder();
+            for (byte aMessageDigest : messageDigest) {
+                String h = Integer.toHexString(0xFF & aMessageDigest);
+                while (h.length() < 2)
+                    h = "0" + h;
+                hexString.append(h);
+            }
+            return hexString.toString();
+        } catch (NoSuchAlgorithmException e) {
+            e.printStackTrace();
```
Comment (Kigamba) [inline src/main/java/com/onaio/steps/helper/HashGenerator.java:30]: Use Timber here

File: src/main/java/com/onaio/steps/helper/AuthDialog.java [hunk 3]
```diff
@@ -229,6 +200,14 @@ public boolean needsAuth() {
         return true;
     }
 
+    public boolean verifyPassword(String enterPwdHash, String storePwdHash, String enterPwd) {
```
Comment (Kigamba) [inline src/main/java/com/onaio/steps/helper/AuthDialog.java:203]: Param names should be `enteredPwdHash`, `storedPwdHash` and `enteredPwd`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java [hunk 4]
```diff
@@ -147,4 +136,45 @@ public void onAuthFailed(AuthDialog authDialog) {
 
     }
 
+    /**
+     * This method test if password is valid in md5 hash update the password
+     * with SHA-256 hash and return true
+     */
+    @Test
+    public void testVerifyPasswordShouldReturnTrueIfMD5HashMatched() {
```
Comment (Kigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:144]: Change to ` testVerifyPasswordShouldReturnTrueIfStoredMD5HashMatched`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java [hunk 5]
```diff
@@ -147,4 +136,45 @@ public void onAuthFailed(AuthDialog authDialog) {
...
+        String text = "test";
+        String md5Hash = "098f6bcd4621d373cade4e832627b4f6";
```
Comment (Kigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:146]: Rename to `storedMd5Hash`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java [hunk 6]
```diff
@@ -147,4 +136,45 @@ public void onAuthFailed(AuthDialog authDialog) {
...
+        String sha256Hash = "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08";
```
Comment (Kigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:147]: Rename to `enteredSha256Hash`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java [hunk 7]
```diff
@@ -147,4 +136,45 @@ public void onAuthFailed(AuthDialog authDialog) {
...
+    }
+
+    /**
+     * This method test if store password and enter passwored is not matched
+     * with SHA-256 hash then return false
+     */
+    @Test
+    public void testVerifyPasswordShouldReturnFalseIfHashNotMatched() {
+        String text = "test";
+        String storePwdHash = "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08";
```
Comment (Kigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:174]: `storedPwdHash`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java [hunk 8]
```diff
@@ -147,4 +136,45 @@ public void onAuthFailed(AuthDialog authDialog) {
...
+        String enterPwdHash = "9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822c"; // invalid hash
```
Comment (Kigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:175]: `enteredPwdHash`

**Review by Kigamba — CHANGES_REQUESTED** [review]

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 5 above)
Comment (ekigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:146]: Rename to `storedMd5Hash`

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 8 above)
Comment (ekigamba) [inline src/test/java/com/onaio/steps/helper/AuthDialogTest.java:175]: `enteredPwdHash`

**Review by ekigamba — CHANGES_REQUESTED** [review]

File: src/main/java/com/onaio/steps/helper/AuthDialog.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/AuthDialog.java]: done.

File: src/main/java/com/onaio/steps/helper/HashGenerator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/HashGenerator.java]: done

File: src/main/java/com/onaio/steps/helper/AuthDialog.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (owais-vd) [reply src/main/java/com/onaio/steps/helper/AuthDialog.java]: done

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (owais-vd) [reply src/test/java/com/onaio/steps/helper/AuthDialogTest.java]: done

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 5 above)
Comment (owais-vd) [reply src/test/java/com/onaio/steps/helper/AuthDialogTest.java]: done

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (owais-vd) [reply src/test/java/com/onaio/steps/helper/AuthDialogTest.java]: done

File: src/test/java/com/onaio/steps/helper/AuthDialogTest.java (hunk 8 above)
Comment (owais-vd) [reply src/test/java/com/onaio/steps/helper/AuthDialogTest.java]: done

**Review by ekigamba — APPROVED** [review]

**Pull request merged**